nasa/harmony-service-lib-py
# Harmony 1125 - Automatically retry temporary request failures

## Jira Issue ID
HARMONY-1125

## Description
Automatically retries temporary request failures using https://urllib3.readthedocs.io/en/stable/reference/urllib3.util.html. 

Here is what the end result should look like on the Harmony side. (Note that I I also tweaked the way 500 download errors are handled as well by giving them a custom exception class so that we don't see the "unknown error" on the Harmony job.)

<img width="1479" alt="Screen Shot 2022-06-02 at 10 05 54 AM" src="https://user-images.githubusercontent.com/34752045/171649043-933b0d5f-5743-45c7-9eba-507c1a76826d.png">

<img width="1306" alt="Screen Shot 2022-06-02 at 10 05 35 AM" src="https://user-images.githubusercontent.com/34752045/171649071-58f11353-6bac-490d-90eb-1724dd8c7772.png">


## Local Test Steps
Checkout harmony-1125 in this repo and in the  Harmony repo. In harmony-service-lib-py activate your virtual environment and run `make install` to ensure the right version of responses is installed before running `make test`.
### Testing with Harmony
1. In harmony-service-lib-py in http.py at the top of the download function set `url='https://httpbin.org/status/502'`.
2. Build a service using your local copy of harmony-service-lib-py (e.g. `make build-image LOCAL_SVCLIB_DIR=../harmony-service-lib-py` in harmony-service-example)
3. In your env file for the harmony repo set `MAX_DOWNLOAD_RETRIES=4` or whatever value you want. 
4. Restart your harmony services in k8s and make a request. The retry history should be logged _after_ the final download failure.

### Testing harmony-service-lib-py only (tedious)
1. Navigate to harmony-service-lib-py root dir.
2. Make sure you have an .env file there with something like:
```
# this is the thing we're testing
MAX_DOWNLOAD_RETRIES=3

OAUTH_CLIENT_ID=123
OAUTH_UID=local_harmony
OAUTH_PASSWORD=abc
OAUTH_REDIRECT_URI=http://localhost:3000/oauth2/redirect
STAGING_PATH=public/harmony/service-example
STAGING_BUCKET=local-staging-bucket
FALLBACK_AUTHN_ENABLED=true
EDL_USERNAME=a
EDL_PASSWORD=b
```
3. Activate your virtual environment and pip install dotenv.
4. In util.py comment out the `download` function and `from harmony import http`. (This will prevent a circular ref later on.)
5. Add the following to the bottom of http.py:
```
import dotenv
from harmony.util import config

print(dotenv.load_dotenv('.env'))
download(config(), 'https://httpbin.org/status/502', None, None, '/file',
             user_agent=None, stream=True, buffer_size=1024*1024*16)
```
6a. (optional) In _download_with_fallback_authn, modify/add this code (which prints the retried requests):
```
resp = session.get(url, headers=headers, timeout=TIMEOUT, **kwargs_download_agent)
for r in resp.raw.retries.history:
    print(r)
return resp
```
6b. (optional) In order to log some of the retry logic from the urllib3 Retry class, navigate to `venv/lib/python3.9/site-packages/urllib3/util/retry.py` (your path may differ). The relevant functions you may want to print things in are `get_backoff_time` and `_sleep_backoff`.
7. From harmony-service-lib-py root dir, run `python -m harmony.http`

With this setup you'll be using the default backoff factor of 2, but that can be modified as well. Similar code can be added to trigger the other http requests in http.py, but it will ultimately produce the same outcome.

## PR Acceptance Checklist
